src/gt4py/gt_src_manager.py

```diff
@@ -28,29 +29,33 @@
 
 
 def install_gt_sources() -> bool:
-    is_ok = has_gt_sources()
-    if not is_ok:
-        GIT_BRANCH = _GRIDTOOLS_GIT_BRANCH
-        GIT_REPO = _GRIDTOOLS_GIT_REPO
+    if has_gt_sources():
+        return True
 
-        install_path = os.path.dirname(__file__)
-        target_path = os.path.abspath(
-            os.path.join(install_path, "_external_src", _GRIDTOOLS_REPO_DIRNAMES)
-        )
-        if not os.path.exists(target_path):
+    GIT_BRANCH = _GRIDTOOLS_GIT_BRANCH
+    GIT_REPO = _GRIDTOOLS_GIT_REPO
+
+    install_path = os.path.dirname(__file__)
+    external_path = os.path.abspath(os.path.join(install_path, "_external_src"))
+    target_path = os.path.join(external_path, _GRIDTOOLS_REPO_DIRNAMES)
+    lock_file = os.path.join(external_path, ".gridtools.lock")
+
+    with open(lock_file, mode="w") as f:
+        fcntl.flock(f.fileno(), fcntl.LOCK_EX)
+        if not has_gt_sources():
             git_cmd = f"git clone --depth 1 -b {GIT_BRANCH} {GIT_REPO} {target_path}"
             print(f"Getting GridTools C++ sources...\n$ {git_cmd}")
             subprocess.check_call(git_cmd.split(), stderr=subprocess.STDOUT)
 
-        is_ok = has_gt_sources()
-        if is_ok:
-            print("Success!!")
-        else:
-            print(
-                f"\nOooops! GridTools sources have not been installed!\n"
-                f"Install them manually in '{install_path}/_external_src/'\n\n"
-                f"\tExample: git clone --depth 1 -b {GIT_BRANCH} {GIT_REPO} {target_path}\n"
-            )
+    is_ok = has_gt_sources()
+    if is_ok:
+        print("Success!!")
+    else:
+        print(
+            f"\nOooops! GridTools sources have not been installed!\n"
+            f"Install them manually in '{install_path}/_external_src/'\n\n"
+            f"\tExample: git clone --depth 1 -b {GIT_BRANCH} {GIT_REPO} {target_path}\n"
+        )
```

Comment:
I know this is the current behavior, but should this really fail silently (without raising an exception)? It was very hard to tell that this was the problem in the past because it's hidden in long log output, and this isn't even logged with ERROR, it's just printed.

Won't this failing always lead to an exception down the line when we try to use the sources?

Reply:
Yes. Nice call out. It's a good opportunity to fix that behavior and raise the proper exception here.

Reply:
Does https://github.com/GridTools/gt4py/pull/938/commits/52ea3517adffa095ca2588691c123006467320b7 look like what you meant?

Reply:
Yes something like this. I would be careful about merging this and broadcast it to the rest of the gt4py team, since this could be very breaking if someone relies on it (even though I don't think anyone should or does).

It also strikes me that because we're always checking whether gt sources exist before calling the install command that there's no need to check whether the sources exist within the install command. In other words, the install command could install them regardless of whether they already exist, and the behavior would be unchanged (as long as the check were also added to the main block to update `python3 -m gt4py.gt_src_manager install`).

A different interpretation could be that we don't need to check if the sources are already installed within the backend code.